### PR TITLE
SDKS-2899 PingOne Protect e2e tests

### PR DESCRIPTION
**Pull request by spetrov**
# JIRA Ticket
[SDKS-2899](https://bugster.forgerock.org/jira/browse/SDKS-2899) [QA][Android] Test Protect SDK

# Description
- Renamed `PingOneProtectInitCallback` class to `PingOneProtectInitializeCallback`.
- Added e2e tests for the PingOneProtectInitialize and PingOneProtectEvaluate callbacks.

### CODE REVIEW

File: ping-protect/src/androidTest/java/org/forgerock/android/auth/BasePingOneProtectTest.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Before
+    fun setUpSDK() {
+        set(Logger.Level.DEBUG)
+        FRAuth.start(context, options)
+    }
+
+    @After
+    fun logoutSession() {
+        if (FRSession.getCurrentSession() != null) {
```
Comment (jeyanthanperiyasamy) [inline ping-protect/src/androidTest/java/org/forgerock/android/auth/BasePingOneProtectTest.kt:55]: FRSession.getCurrentSession()?.let {
  it.logout()
}

File: ping-protect/src/androidTest/java/org/forgerock/android/auth/BasePingOneProtectTest.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2023 ForgeRock. All rights reserved.
```
Comment (jeyanthanperiyasamy) [inline ping-protect/src/androidTest/java/org/forgerock/android/auth/BasePingOneProtectTest.kt:2]: copyright 2024

File: ping-protect/src/androidTest/java/org/forgerock/android/auth/BasePingOneProtectTest.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import org.forgerock.android.auth.Logger.Companion.set
+import org.junit.After
+import org.junit.Before
+import org.junit.BeforeClass
+import org.junit.Rule
+import org.junit.rules.Timeout
+import java.util.concurrent.TimeUnit
+
+abstract class BasePingOneProtectTest {
+    var context: Context = ApplicationProvider.getApplicationContext()
```
Comment (jeyanthanperiyasamy) [inline ping-protect/src/androidTest/java/org/forgerock/android/auth/BasePingOneProtectTest.kt:20]: val context: Context = ApplicationProvider.getApplicationContext()

File: ping-protect/src/androidTest/java/org/forgerock/android/auth/BasePingOneProtectTest.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ * of the MIT license. See the LICENSE file for details.
+ */
+package org.forgerock.android.auth
+
+import android.content.Context
+import androidx.test.core.app.ApplicationProvider
+import org.forgerock.android.auth.Logger.Companion.set
+import org.junit.After
+import org.junit.Before
+import org.junit.BeforeClass
```
Comment (jeyanthanperiyasamy) [inline ping-protect/src/androidTest/java/org/forgerock/android/auth/BasePingOneProtectTest.kt:14]: BeforeClass remove , its not used

File: ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectEvaluateCallbackTest.kt [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+import org.junit.Assert
+import org.junit.FixMethodOrder
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.MethodSorters
+
+@RunWith(AndroidJUnit4::class)
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+class PingOneProtectEvaluateCallbackTest : BasePingOneProtectTest() {
+    private val TREE = "TEST_PING_ONE_PROTECT_EVALUATE"
```
Comment (jeyanthanperiyasamy) [inline ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectEvaluateCallbackTest.kt:25]: private val tree = "TEST_PING_ONE_PROTECT_EVALUATE"

property variable names are lowerCamelCase
const are UPPERCASE

File: ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectEvaluateCallbackTest.kt [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    @Test
+    fun test01EvaluateNoInit() {
+        var evaluateFailure = false
+        val nodeListenerFuture: PingOneProtectNodeListener = object : PingOneProtectNodeListener(
+            context, "evaluate-no-init"
+        ) {
+            val nodeListener: NodeListener<FRSession?> = this
+            override fun onCallbackReceived(node: Node) {
+
+                if (node.getCallback(PingOneProtectEvaluationCallback::class.java) != null) {
```
Comment (jeyanthanperiyasamy) [inline ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectEvaluateCallbackTest.kt:36]: i think we can simplify this 

`   override fun onCallbackReceived(node: Node) {
                
                node.getCallback(PingOneProtectEvaluationCallback::class.java)?.let {
                    val scope = CoroutineScope(Dispatchers.Default)
                    scope.launch {
                        try {
                            it.getData(context)
                        }
                        catch (e: Exception) {
                            evaluateFailure = true
                            assertThat("PingOneSignals SDK is not initialized", e.message === "PingOneSignals SDK is not initialized")
                        }
                        node.next(context, nodeListener)
                    }
                    return
                }

                node.getCallback(TextOutputCallback::class.java)?.let {
                    val callback = node.getCallback(
                        TextOutputCallback::class.java
                    )

                    assertThat("The Protect Evaluation node should trigger the Client Error outcome", callback.message == "Client Error")
                    node.next(context, nodeListener)
                    return
                }
                
                super.onCallbackReceived(node)
            }`

File: ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectNodeListener.kt [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import android.content.Context
+import org.forgerock.android.auth.callback.ChoiceCallback
+import org.forgerock.android.auth.callback.NameCallback
+
+open class PingOneProtectNodeListener(
+    private val context: Context,
+    private val nodeConfiguration: String
+) : NodeListenerFuture<FRSession?>() {
+    override fun onCallbackReceived(node: Node) {
+        if (node.getCallback(ChoiceCallback::class.java) != null) {
```
Comment (jeyanthanperiyasamy) [inline ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectNodeListener.kt:18]:
```
  node.getCallback(ChoiceCallback::class.java)?.let {
            val choices = it.choices
            val choiceIndex = choices.indexOf(nodeConfiguration)
            it.setSelectedIndex(choiceIndex)
            node.next(context, this)
        }

        node.getCallback(NameCallback::class.java)?.let {
            it.setName(BasePingOneProtectTest.USERNAME)
            node.next(context, this)
        }
        

```

**Review by jeyanthanperiyasamy — CHANGES_REQUESTED** [review]
gave comments

File: ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectEvaluateCallbackTest.kt (hunk 5 above)
Comment (spetrov) [reply ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectEvaluateCallbackTest.kt:25]: Refactored this to:
`private val authenticationTree = "TEST_PING_ONE_PROTECT_EVALUATE"`

File: ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectInitializeCallbackTest.kt [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import org.forgerock.android.auth.callback.TextOutputCallback
+import org.hamcrest.MatcherAssert.assertThat
+import org.junit.Assert
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class PingOneProtectInitializeCallbackTest : BasePingOneProtectTest() {
```
[the remark this answers is not in the document]
Comment (spetrov) [reply ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectInitializeCallbackTest.kt:21]: Refactored...

File: ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectEvaluateCallbackTest.kt (hunk 6 above)
Comment (spetrov) [reply ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectEvaluateCallbackTest.kt:36]: Thanks! I applied the suggestion to all test cases...

File: ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectNodeListener.kt (hunk 7 above)
Comment (spetrov) [reply ping-protect/src/androidTest/java/org/forgerock/android/auth/PingOneProtectNodeListener.kt:18]: Refactored as suggested! Thanks!

**Review by jeyanthanperiyasamy — APPROVED** [review]

File: forgerock-integration-tests/src/androidTest/java/org/forgerock/android/auth/ProtectCallbackTest.kt [hunk 9]
```diff
@@ -42,8 +42,8 @@ class ProtectCallbackTest {
         ) {
             val nodeListener: NodeListener<FRSession?> = this
             override fun onCallbackReceived(node: Node) {
-                if (node.getCallback<PingOneProtectInitCallback>(PingOneProtectInitCallback::class.java) != null) {
-                    val callback: PingOneProtectInitCallback = node.getCallback<PingOneProtectInitCallback>(PingOneProtectInitCallback::class.java)
+                if (node.getCallback<PingOneProtectInitializeCallback>(PingOneProtectInitializeCallback::class.java) != null) {
```
Comment (jeyanthanperiyasamy) [inline forgerock-integration-tests/src/androidTest/java/org/forgerock/android/auth/ProtectCallbackTest.kt:45]: (node.getCallback<PingOneProtectInitializeCallback>(PingOneProtectInitializeCallback::class.java)?.let {

}

just one update here as well ...all good

Comment (spetrov) [reply forgerock-integration-tests/src/androidTest/java/org/forgerock/android/auth/ProtectCallbackTest.kt]: Oh, this file actually has to be removed...

**Pull request merged**